### PR TITLE
nft: Open a persistent connection to mnl/netlink socket at init_redirect()

### PR DESCRIPTION
**Pull request by paul-chambers**
Preparation for dropping elevated privileges after the initialization phase for the nftables implementation on Linux. See #405 

Move the 8k allocation on the stack for debug-printing rules in debug builds into a separate function. This keeps the stack layout consistent between debug and non-debug builds, making it easier to diagnose some classes of bugs, e.g. stack corruptions, alignment problems, etc.

### CODE REVIEW

File: miniupnpd/netfilter_nft/nftnlrdr_misc.c [hunk 1]
```diff
@@ -82,7 +83,37 @@ static uint32_t rule_list_filter_validate = RULE_CACHE_INVALID;
...
+void
+nft_mnl_disconnect(void)
+{
+	mnl_socket_close(mnl_sock);
+	mnl_sock = NULL;
+}
+
 #ifdef DEBUG
+
+#if 0
```
Comment (miniupnp) [inline miniupnpd/netfilter_nft/nftnlrdr_misc.c]: maybe you should comment why this code is left dead ;)

**Comment by miniupnp** [comment]
thanks !

**Comment by miniupnp** [comment]
thank you for investing time in miniupnpd.
I'm quite busy theses days and I don't have much time to setup a router with nftables to fix and test the nftables backend

**Comment by paul-chambers** [comment]
Thanks for creating miniupnp in the first place. I've been using it for years, seems only fair to contribute back when the chance presented itself :)

Perhaps once I've proven myself, I can help out more. I'd love to help separate the single repo into per-project repos for example ( #152 ). My OCD again :)

The nftables implementation could be made simpler. Not sure if the caches are really necessary, the nfnml library provides iterators that could serve the same purposes, and I'm not sure the performance gain is worth the major increase in complexity. The rule_t private structure to mirror the opaque struct nftnl_rule has to be (re)populated by multiple API calls anyhow.

File: miniupnpd/netfilter_nft/nftnlrdr_misc.c (hunk 1 above)
Comment (paul-chambers) [reply miniupnpd/netfilter_nft/nftnlrdr_misc.c]: I just removed it, it's been dead for a while :)  I just tend to tread lightly while I'm still getting up to speed on someone else's code.

**Comment by paul-chambers** [comment]
Hmm... I'm confused by the conflicts. They don't actually look like conflicts to me.

**Comment by miniupnp** [comment]
I will fix the conflicts
I have the bad habit of rewriting the history ;)

**Comment by miniupnp** [comment]
OK. I merged myself.
@paul-chambers I think you need to have a look at _nftpinhole.c_
please build with --igd2 and --ipv6 options

**Comment by paul-chambers** [comment]
Thank you for cleaning up after me - sorry about that. The dangers of coding when tired. I'll get travis installed too.

Would it make sense to make those two options configurable by .conf, rather than at compile time?

**Comment by miniupnp** [comment]
There is no reason to build theses features for them being disabled.

**Pull request closed without merging**